### ISSUE TITLE
Is there a way to declare a non primary key auto increment field?

### ISSUE THREAD

**Issue by ali-web** [typeorm/typeorm#1517]
Both MySQL and Postgres allow this AFAIK, but I wonder if typeorm supports this.

**Comment by daniel-lang** [comment]
No, TypeORM doesn't support auto-increment columns that aren't primary keys. But it shouldn't be too hard to add it.
@pleerock is that something you would be interested in?

**Comment by gcristea07** [comment]
@Generated('increment')
    id: number;

**Comment by taisuke-j** [comment]
```
@Column()
@Generated('increment')
id: number;
```
If you also need to get the field

**Comment by Ollie1700** [comment]
For anyone coming across this problem with Postgres - and specifically needing to do this as part of a migration - this worked for our team:

```typescript
import {MigrationInterface, QueryRunner, TableColumn} from "typeorm";

export class AddOrderToEvent1611919059696 implements MigrationInterface {

    public async up(queryRunner: QueryRunner): Promise<void> {
      await queryRunner.addColumn('event', new TableColumn({
        name: 'order',
        // "serial" type is equivalent to AI in PG
        type: 'serial',
        isNullable: true
      }))
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
      await queryRunner.dropColumn('event', 'order')
    }

}
```

**Comment by Oscarclarosc** [comment]
First define the property as Unique
`@Unique(["propertyName"])`

...inside class entity
```
@Column()
@Generated("increment")
propertyName: number;
```

**Comment by pratiksyngenta** [comment]
Its giving me
`Incorrect table definition; there can be only one auto column and it must be defined as a key` error

**Comment by eomttt** [comment]
> 

so do i

**Comment by aaliyan1230** [comment]
is there a way to custom auto generate string fields like this ?? for instance, generating custom string 4 digit ticket tracking IDs of form 'Txxxx' where xxxx are numbers to be present in the ID

**Comment by Phoenix-Alpha** [comment]
> is there a way to custom auto generate string fields like this ?? for instance, generating custom string 4 digit ticket tracking IDs of form 'Txxxx' where xxxx are numbers to be present in the ID

https://wanago.io/2021/11/29/generated-columns-postgresql-typeorm/

**Comment by LHoOXD** [comment]
> Its giving me `Incorrect table definition; there can be only one auto column and it must be defined as a key` error

if you run to this problem Then add the @Index decorator , like this : 
@Index()
@Column()
@Generated("increment")
propertyName: number;

**Comment by tinnguyen69** [comment]
For those using MySQL and defining your column like this
> ```
> @Column()
> @Generated('increment')
> id: number;
> ```

and get the error below
> Its giving me `Incorrect table definition; there can be only one auto column and it must be defined as a key` error

You will need to manually change your migration file.
When you create the migration, it will look like this:
`
    await queryRunner.query(`ALTER TABLE \`xxx\` ADD \`id\` int NOT NULL AUTO_INCREMENT`);
    await queryRunner.query(
      `CREATE INDEX \`IDX_329b07a62ca0e5160963459e30\` ON \`xxx\` (\`id\`)`,
    );
`

change it to
`
    await queryRunner.query(`ALTER TABLE \`xxx\` ADD \`id\` int`);
    await queryRunner.query(
      `CREATE INDEX \`IDX_329b07a62ca0e5160963459e30\` ON \`xxx\` (\`id\`)`,
    );
    await queryRunner.query(
      `ALTER TABLE \`xxx\` CHANGE \`id\` \`search_index\` int NOT NULL AUTO_INCREMENT`,
    );
`

Hope it help.
